### PR TITLE
Switch to reason

### PR DESCRIPTION
**Pull request by johnhaley81**

### CODE REVIEW

File: src/zafClient.re [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+      "_type": string
+    }
+  ) =>
+  Js.Promise.t(Js.Json.t) =
+  "";
+
+[@bs.send]
+external on_app_registered :
+  (zafClient, [@bs.as "app.registered"] _, unit => unit) => unit =
+  "on";
```
Comment (smith-kyle) [inline src/zafClient.re]: can you break this down for me?

What I get from it so far is "on_app_registered is a function that accepts a zafClient, a function that goes from unit => unit, and returns a unit. Also it has hard coded 'app.registered' as the first argument to zafClient.on". Is that all right? so it compiles to

```javascript
const on_app_registred = (x, y) => x.on('app.registered', y);
```

but my real question is why the `_` after `[@bs.as "app.registered"]`

File: src/zafClient.re [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    |> then_(json => json |> Decode.context |> resolve)
+  );
+
+let accountLens =
+  Rationale.Lens.(
+    make(
+      context => context.account,
+      (account, context) => {...context, account}
+    )
+    >>- optional({subdomain: None})
```
Comment (smith-kyle) [inline src/zafClient.re]: It strikes me as odd that we have to use a different pipe function for Rationale

File: src/zafClient.re (hunk 1 above)
Comment (johnhaley81) [reply src/zafClient.re]: I lifted it from https://bucklescript.github.io/docs/en/function.html#fixed-arguments

I don't actually know what the `_` does :)

File: src/zafClient.re (hunk 2 above)
Comment (johnhaley81) [reply src/zafClient.re]: So to do things in a monad you have to have something "unwrap it". Like with a `Promise` you don't just pipe it, you pipe it to `then_` which unwraps the `Promise` and allows you to do something to the type inside of it.

This is how you have to chain monads if you want to work with the value inside of it. We can pipe a monad to something else but we have to use a different operator to access the value inside.

What this is doing is removing the monad and giving us a final value of either `Some(value)` or `{subdomain: None}` which we get from https://github.com/jonlaing/rationale/blob/master/src/Lens.re#L26

**Pull request merged**